auth/auth.go

```diff
@@ -495,6 +497,9 @@ type Options2LO struct {
 	// enabled by setting GOOGLE_SDK_GO_LOGGING_LEVEL in which case a default
 	// logger will be used. Optional.
 	Logger *slog.Logger
+	// TokenHook is a function that will be called after a token is fetched,
+	// allowing for modification of the token.
+	TokenHook TokenHook
```

Comment:
One more idea. I think we can refactor the idea of having the token hook public away as well to make virtual no changes to the public API for this feature.
1. Move the detection logic of your feature to your internal package.
2. Wrap it in a sync.Once. If you want an example see: https://github.com/googleapis/google-cloud-go/blob/7ec4dd4e8318387d9998642babf3715ec790f728/compute/metadata/metadata.go#L127
3. When the TokenProviders are being created instead insert your hook just as you do now, but only if the feature is enabled.

Thoughts?

Reply:
Good idea! Removed TokenHook and replaced it with a provider that wraps the base TokenProvider.